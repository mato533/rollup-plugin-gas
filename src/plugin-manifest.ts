import { existsSync, readFileSync } from "fs";
import { join } from "path";
import type { Plugin } from "rollup";
import pc from "picocolors";
import type { NotNullRollupPluginGasOptions } from "types";
import { getRelativePath } from "@/plugin-utils";

export const manifest = "appsscript.json";

export const loadManifest = (sourceFile: string) => {
  return readFileSync(sourceFile, {
    encoding: "utf8",
  });
};

const rollupPluginGasCopyManifest = (
  configuratedOptions: NotNullRollupPluginGasOptions
): Plugin => {
  return {
    name: "rollup-plugin-gas-copy-manifest",
    generateBundle() {
      if (!configuratedOptions.manifest.copy) {
        return;
      }
      const sourceFile = join(configuratedOptions.manifest.srcDir, manifest);
      this.info(
        pc.gray("Copy the manifest from: ") +
          pc.green(getRelativePath(sourceFile))
      );

      if (!existsSync(sourceFile)) {
        this.error("Manifest file is not exist: " + sourceFile);
      }
      this.emitFile({
        type: "asset",
        name: "manifest",
        fileName: manifest,
        source: loadManifest(sourceFile),
      });
    },
  };
};

export default rollupPluginGasCopyManifest;
