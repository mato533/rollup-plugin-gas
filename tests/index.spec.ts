import path from "path";
import pc from "picocolors";
import { rollup } from "rollup";
import type { OutputAsset } from "rollup";
import rollupPluginGas from "../src";
import type { RollupPluginGasOptions } from "types";
import { manifest } from "@/plugin-manifest";

interface TestParams {
  scenario: string;
  dirFixtures: string;
}

const dirFixtures = path.resolve(__dirname, "./__fixtures__");
const dirIncludeFixtures = path.resolve(dirFixtures, "include");

const defineFixtureFileName = (param: TestParams) => {
  const { scenario, dirFixtures: dir } = param;
  const inputFile = path.join(dir, `${scenario}.js`);
  return inputFile;
};

const build = async (inputFile: string, options?: RollupPluginGasOptions) => {
  const bundle = await rollup({
    input: inputFile,
    plugins: [rollupPluginGas(options)],
  });

  return await bundle.generate({});
};

const buildAndAssertOutput = async (
  param: TestParams,
  options?: RollupPluginGasOptions
) => {
  const inputFile = defineFixtureFileName(param);

  const output = await build(inputFile, options);

  expect(output.output.length).toBe(1);
  const [{ code: generated }] = output.output;
  expect(generated).toMatchSnapshot();
};

const buildAndAssertManifest = async (option: RollupPluginGasOptions) => {
  const inputFile = defineFixtureFileName({
    scenario: "basic",
    dirFixtures: dirFixtures,
  });

  const output = await build(inputFile, option);
  expect(output.output.length).toBe(2);
  const generatedManifest = output.output[1] as OutputAsset;
  expect(generatedManifest.type).toBe("asset");
  expect(generatedManifest.fileName).toBe(manifest);
  expect(generatedManifest.source).toMatchSnapshot();
};

describe("rollup-plugin-gas", () => {
  it.each(["basic", "allow-function"])(
    "Should add entry point function to the bandle file #%s",
    async (scenario) => {
      await buildAndAssertOutput({
        scenario: scenario,
        dirFixtures: dirFixtures,
      });
    }
  );

  it("Should include comments", async () => {
    await buildAndAssertOutput(
      { scenario: "comment", dirFixtures: dirFixtures },
      { comment: true }
    );
  });

  it("Should add entrypoint function of the specified file", async () => {
    await buildAndAssertOutput(
      { scenario: "include", dirFixtures: dirIncludeFixtures },
      { include: ["**/include.js"] }
    );
  });

  it("Should print each source filename to bandle file", async () => {
    await buildAndAssertOutput(
      {
        scenario: "comment-module",
        dirFixtures: dirFixtures,
      },
      { moduleHeaderComment: true }
    );
  });

  it("Should print each source filename to bandle file", async () => {
    await buildAndAssertOutput(
      {
        scenario: "include-comment-module",
        dirFixtures: dirIncludeFixtures,
      },
      { include: ["**/include-comment-module.js"], moduleHeaderComment: true }
    );
  });

  it("Should copy the manifest file", async () => {
    vi.spyOn(process, "cwd").mockReturnValue(dirFixtures);

    await buildAndAssertManifest({
      manifest: { copy: true },
    });
  });

  it("Should copy the manifest file using srcDir option", async () => {
    vi.spyOn(process, "cwd").mockReturnValue(__dirname);

    await buildAndAssertManifest({
      manifest: { copy: true, srcDir: "./__fixtures__" },
    });
  });
});
