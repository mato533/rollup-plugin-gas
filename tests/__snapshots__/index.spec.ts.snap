// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rollup-plugin-gas > Should add entry point function to the bandle file #allow-function 1`] = `
"
var global = this;
function basic() {};
(function (factory) {
  typeof define === 'function' && define.amd ? define(factory) :
  factory();
})((function () { 'use strict';

  global.basic = () => {
    console.log(\\"test\\");
  };

}));
"
`;

exports[`rollup-plugin-gas > Should add entry point function to the bandle file #basic 1`] = `
"
var global = this;
function echo() {};
function hello() {};
function plus() {};
function minus() {};
(function (factory) {
  typeof define === 'function' && define.amd ? define(factory) :
  factory();
})((function () { 'use strict';

  const echo = (message) => {
    return message;
  };

  /**
   * Return write arguments.
   */
  global.echo = echo;

  global.hello = function () {
    console.log(\\"hello\\");
  };

  function plus(x, y) {
    return x + y;
  }
  function minus(x, y) {
    return x - y;
  }
  (global.plus = plus), (global.minus = minus);

}));
"
`;

exports[`rollup-plugin-gas > Should add entrypoint function of the specified file 1`] = `
"
var global = this;
function includeTest() {};
(function (factory) {
  typeof define === 'function' && define.amd ? define(factory) :
  factory();
})((function () { 'use strict';

  global.excludeFunction = () => {
    console.log(\\"excludeFunction\\");
  };

  const foo = function () {
    console.log(\\"foo\\");
  };

  global.includeTest = function () {
    console.log(\\"includeTest\\");
    foo();
  };

}));
"
`;

exports[`rollup-plugin-gas > Should copy the manifest file 1`] = `
"{
  \\"timeZone\\": \\"Asia/Tokyo\\",
  \\"dependencies\\": {},
  \\"webapp\\": {
    \\"access\\": \\"MYSELF\\",
    \\"executeAs\\": \\"USER_ACCESSING\\"
  },
  \\"exceptionLogging\\": \\"STACKDRIVER\\",
  \\"runtimeVersion\\": \\"V8\\"
}
"
`;

exports[`rollup-plugin-gas > Should copy the manifest file using srcDir option 1`] = `
"{
  \\"timeZone\\": \\"Asia/Tokyo\\",
  \\"dependencies\\": {},
  \\"webapp\\": {
    \\"access\\": \\"MYSELF\\",
    \\"executeAs\\": \\"USER_ACCESSING\\"
  },
  \\"exceptionLogging\\": \\"STACKDRIVER\\",
  \\"runtimeVersion\\": \\"V8\\"
}
"
`;

exports[`rollup-plugin-gas > Should include comments 1`] = `
"
var global = this;
/**
 * Comment for echo
 */
function echo() {};
/**
 * Comment foo
 */
function foo(message) {};
(function (factory) {
  typeof define === 'function' && define.amd ? define(factory) :
  factory();
})((function () { 'use strict';

  const echo = (message) => {
    return message;
  };

  // header comment
  /**
   * Comment for echo
   */
  global.echo = echo;

  /**
   * Comment foo
   */
  global.foo = (message) => {
    // comment in the logic
    console.log(message);
  };
  // end Comment

}));
"
`;

exports[`rollup-plugin-gas > Should print each source filename to bandle file 1`] = `
"
var global = this;
function echo() {};
function foo(message) {};
(function (factory) {
  typeof define === 'function' && define.amd ? define(factory) :
  factory();
})((function () { 'use strict';

  /*!***************!*\\\\
    !*** echo.js ***!
  \\\\*!***************!*/
  const echo = (message) => {
    return message;
  };

  /*!*************************!*\\\\
    !*** comment-module.js ***!
  \\\\*!*************************!*/
  // header comment
  /**
   * Comment for echo
   */
  global.echo = echo;

  /**
   * Comment foo
   */
  global.foo = (message) => {
    // comment in the logic
    console.log(message);
  };
  // end Comment

}));
"
`;

exports[`rollup-plugin-gas > Should print each source filename to bandle file 2`] = `
"
var global = this;
function includeTest() {};
(function (factory) {
  typeof define === 'function' && define.amd ? define(factory) :
  factory();
})((function () { 'use strict';

  /*!**************!*\\\\
    !*** foo.js ***!
  \\\\*!**************!*/
  global.excludeFunction = () => {
    console.log(\\"excludeFunction\\");
  };

  const foo = function () {
    console.log(\\"foo\\");
  };

  /*!*********************************!*\\\\
    !*** include-comment-module.js ***!
  \\\\*!*********************************!*/

  global.includeTest = function () {
    console.log(\\"includeTest\\");
    foo();
  };

}));
"
`;
